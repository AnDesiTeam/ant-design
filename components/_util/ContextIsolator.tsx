import React from 'react';

import { NoFormStyle } from '../form/context';
import { NoCompactStyle } from '../space/Compact';

const ContextIsolator: React.FC<Readonly<React.PropsWithChildren>> = ({ children }) => (
  <NoCompactStyle>
    <NoFormStyle override status>
      {children}
    </NoFormStyle>
  </NoCompactStyle>
);

if (process.env.NODE_ENV !== 'production') {
  ContextIsolator.displayName = 'ContextIsolator';
}

export default ContextIsolator;
