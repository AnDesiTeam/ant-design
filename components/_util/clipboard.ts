// eslint-disable-next-line import/prefer-default-export
export const writeText = (content: string, format?: 'text/plain' | 'text/html') =>
  new Promise<void>((resolve, reject) => {
    const range = document.createRange();
    range.selectNodeContents(document.body);
    document.getSelection()?.addRange(range);

    let success = false;
    const listener = (e: ClipboardEvent) => {
      e.preventDefault();
      const clipboardData = e.clipboardData || (window as any).clipboardData;
      clipboardData?.clearData();
      clipboardData?.setData(format || 'text/plain', content);
      success = true;
    };
    document.addEventListener('copy', listener);
    document.execCommand('copy');
    document.removeEventListener('copy', listener);
    document.getSelection()?.removeAllRanges();
    if (success) {
      resolve();
    } else {
      reject();
    }
  });
