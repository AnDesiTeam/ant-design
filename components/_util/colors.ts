import type { ElementOf } from './type';

export const PresetStatusColorTypes = [
  'success',
  'processing',
  'error',
  'default',
  'warning',
] as const;
// eslint-disable-next-line import/prefer-default-export
export const PresetColorTypes = [
  'pink',
  'red',
  'yellow',
  'orange',
  'cyan',
  'green',
  'blue',
  'purple',
  'geekblue',
  'magenta',
  'volcano',
  'gold',
  'lime',
] as const;

export type PresetColorType = ElementOf<typeof PresetColorTypes>;
export type PresetStatusColorType = ElementOf<typeof PresetStatusColorTypes>;
