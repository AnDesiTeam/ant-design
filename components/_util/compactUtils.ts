/* eslint-disable import/prefer-default-export */
import classNames from 'classnames';

import type { SpaceCompactItemContextType } from '../space/Compact';

export function getCompactClassNames(
  prefixCls: string,
  compactItemContext: Omit<SpaceCompactItemContextType, 'size'> = {},
) {
  const { direction, isItem, isFirstItem, isLastItem } = compactItemContext;
  const joiner = direction === 'vertical' ? 'vertical-' : '-';

  return classNames({
    [`${prefixCls}-compact${joiner}item`]: isItem,
    [`${prefixCls}-compact${joiner}first-item`]: isFirstItem,
    [`${prefixCls}-compact${joiner}last-item`]: isLastItem,
  });
}
