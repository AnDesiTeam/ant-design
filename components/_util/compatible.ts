/* eslint-disable import/prefer-default-export */
import * as React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { createRoot, Root } from 'react-dom/client';

const MARK = '__antd_react_root__';

export function reactRender(node: React.ReactElement, container: Element | DocumentFragment) {
  if (createRoot !== undefined) {
    const root: Root = (container as any)[MARK] || createRoot(container);
    root.render(node);

    (container as any)[MARK] = root;
    return;
  }

  render(node, container);
}

export function reactUnmount(container: Element | DocumentFragment) {
  if (createRoot !== undefined) {
    // Delay to unmount to avoid React 18 sync warning
    Promise.resolve().then(() => {
      (container as any)[MARK]?.unmount();
    });

    return;
  }

  unmountComponentAtNode(container);
}
