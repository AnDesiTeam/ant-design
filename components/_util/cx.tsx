import classNames from 'classnames';

// space
const s = ' ';

export default (prefixCls: string) => {
  const cx = (...params: any) => {
    const str = classNames(...params);
    const arr = str.split(s);
    return arr.map(item => `${prefixCls}-${item}`).join(s);
  };

  const concatCls = (str: string) => {
    return (...params: any) => {
      const pre = str === '' ? '' : `${str}${s}`;

      return `${pre}${cx(...params)}`;
    };
  };

  const mixUnprefix = (...params: any) => {
    const str = classNames(...params);
    return concatCls(str);
  };

  const skipFirst = (...params: any) => {
    const [first, ...rest] = params;
    return mixUnprefix(first)(...rest);
  };

  const cxRoot = (...params: any) => {
    return `${prefixCls}${s}${cx(...params)}`;
  };

  return {
    classNames,
    cx,
    cxRoot,
    // alias
    cxsf: skipFirst,
  };
};
