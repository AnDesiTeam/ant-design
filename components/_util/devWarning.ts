import devWarning, { resetWarned, noteOnce } from 'rc-util/lib/warning';

export { resetWarned };

export default (valid: boolean, component: string, message: string): void => {
  devWarning(valid, `[antd: ${component}] ${message}`);
};

export const devNote = (valid: boolean, component: string, message: string): void => {
  noteOnce(valid, `[antd: ${component}] ${message}`);
};
