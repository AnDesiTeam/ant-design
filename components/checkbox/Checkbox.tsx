import * as React from 'react';
import classNames from 'classnames';
import type {
  CheckboxRef,
  CheckboxProps as RcCheckboxProps,
  CheckboxChangeEvent,
} from 'rc-checkbox';
import RcCheckbox from 'rc-checkbox';

import Wave from '../_util/wave';
import { TARGET_CLS } from '../_util/wave/interface';
import { ConfigContext } from '../config-provider';
import DisabledContext from '../config-provider/DisabledContext';
import { FormItemInputContext } from '../form/context';
import GroupContext from './GroupContext';
import useStyle from './style';
import useMergedState from 'rc-util/lib/hooks/useMergedState';

export interface AbstractCheckboxProps extends Pick<RcCheckboxProps, 'onChange'> {
  prefixCls?: string;
  className?: string;
  rootClassName?: string;
  defaultChecked?: boolean;
  checked?: boolean;
  /**
   * Alias for `defaultChecked`.
   * @since 5.12.0
   */
  defaultValue?: boolean;
  style?: React.CSSProperties;
  disabled?: boolean;
  title?: string;
  onClick?: React.MouseEventHandler<HTMLElement>;
  onMouseEnter?: React.MouseEventHandler<HTMLElement>;
  onMouseLeave?: React.MouseEventHandler<HTMLElement>;
  onKeyPress?: React.KeyboardEventHandler<HTMLElement>;
  onKeyDown?: React.KeyboardEventHandler<HTMLElement>;
  value?: any; // (native input value) | boolean
  tabIndex?: number;
  name?: string;
  children?: React.ReactNode;
  id?: string;
  autoFocus?: boolean;
  type?: string;
  skipGroup?: boolean;
}

export interface CheckboxChangeEventTarget extends CheckboxProps {
  checked: boolean;
}

export type { CheckboxChangeEvent };

export interface CheckboxProps extends AbstractCheckboxProps {
  indeterminate?: boolean;
}

const InternalCheckbox: React.ForwardRefRenderFunction<CheckboxRef, CheckboxProps> = (
  props,
  ref,
) => {
  const {
    prefixCls: customizePrefixCls,
    className,
    rootClassName,
    children,
    indeterminate = false,
    style,
    onMouseEnter,
    onMouseLeave,
    skipGroup = false,
    disabled,
    checked: checkedProp,
    defaultChecked: defaultCheckedProp,
    value,
    defaultValue,
    ...restProps
  } = props;
  const { getPrefixCls, direction, checkbox } = React.useContext(ConfigContext);
  const checkboxGroup = React.useContext(GroupContext);
  const { isFormItemInput } = React.useContext(FormItemInputContext);
  const contextDisabled = React.useContext(DisabledContext);
  const mergedDisabled = (checkboxGroup?.disabled || disabled) ?? contextDisabled;

  const [checked, setChecked] = useMergedState<boolean>(false, {
    value:
      checkedProp ??
      /**
       * The native html input element does not allow passing a value of bool.
       * ooh, this is a bad hack, but it works.
       */
      (checkboxGroup || typeof value !== 'boolean' ? undefined : value),
    defaultValue: defaultCheckedProp ?? defaultValue,
  });

  const prevValue = React.useRef(value);

  React.useEffect(() => {
    checkboxGroup?.registerValue(value);
  }, []);

  React.useEffect(() => {
    if (skipGroup) {
      return;
    }
    if (value !== prevValue.current) {
      checkboxGroup?.cancelValue(prevValue.current);
      checkboxGroup?.registerValue(value);
      prevValue.current = value;
    }
    return () => checkboxGroup?.cancelValue(value);
  }, [value]);

  const prefixCls = getPrefixCls('checkbox', customizePrefixCls);
  const [wrapSSR, hashId] = useStyle(prefixCls);

  const changeHandler = (event: CheckboxChangeEvent) => {
    restProps.onChange?.(event);
    setChecked(event.target.checked);
  };

  const checkboxProps: Omit<CheckboxProps, 'defaultValue'> = {
    ...restProps,
    checked,
    onChange: changeHandler,
    // bad hack :(, because rc-checkbox passes directly to the html input tag.
    ...(value ? { value } : {}),
  };

  if (checkboxGroup && !skipGroup) {
    checkboxProps.onChange = (...args) => {
      restProps.onChange?.(...args);
      checkboxGroup.toggleOption?.({ label: children, value });
    };
    checkboxProps.name = checkboxGroup.name;
    checkboxProps.checked = checkboxGroup.value.includes(value);
  }

  const classString = classNames(
    `${prefixCls}-wrapper`,
    {
      [`${prefixCls}-rtl`]: direction === 'rtl',
      [`${prefixCls}-wrapper-checked`]: checkboxProps.checked,
      [`${prefixCls}-wrapper-disabled`]: mergedDisabled,
      [`${prefixCls}-wrapper-in-form-item`]: isFormItemInput,
    },
    checkbox?.className,
    className,
    rootClassName,
    hashId,
  );
  const checkboxClass = classNames(
    {
      [`${prefixCls}-indeterminate`]: indeterminate,
    },
    TARGET_CLS,
    hashId,
  );
  const ariaChecked = indeterminate ? 'mixed' : undefined;
  return wrapSSR(
    <Wave component="Checkbox" disabled={mergedDisabled}>
      {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
      <label
        className={classString}
        style={{ ...checkbox?.style, ...style }}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
      >
        <RcCheckbox
          aria-checked={ariaChecked}
          {...checkboxProps}
          prefixCls={prefixCls}
          className={checkboxClass}
          disabled={mergedDisabled}
          ref={ref}
        />
        {children !== undefined && <span>{children}</span>}
      </label>
    </Wave>,
  );
};

const Checkbox = React.forwardRef<CheckboxRef, CheckboxProps>(InternalCheckbox);

if (process.env.NODE_ENV !== 'production') {
  Checkbox.displayName = 'Checkbox';
}

export default Checkbox;
