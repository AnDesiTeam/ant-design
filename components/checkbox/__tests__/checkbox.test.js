import React from 'react';
import { mount } from 'enzyme';
import Checkbox from '..';
import focusTest from '../../../tests/shared/focusTest';
import { resetWarned } from '../../_util/devWarning';
import mountTest from '../../../tests/shared/mountTest';
import rtlTest from '../../../tests/shared/rtlTest';

describe('Checkbox', () => {
  focusTest(Checkbox, { refFocus: true });
  mountTest(Checkbox);
  rtlTest(Checkbox);

  it('responses hover events', () => {
    const onMouseEnter = jest.fn();
    const onMouseLeave = jest.fn();

    const wrapper = mount(<Checkbox onMouseEnter={onMouseEnter} onMouseLeave={onMouseLeave} />);

    wrapper.find('label').simulate('mouseenter');
    expect(onMouseEnter).toHaveBeenCalled();

    wrapper.find('label').simulate('mouseleave');
    expect(onMouseLeave).toHaveBeenCalled();
  });

  it('warning if set `value`', () => {
    resetWarned();

    const errorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
    mount(<Checkbox value />);
    expect(errorSpy).toHaveBeenCalledWith(
      'Warning: [antd: Checkbox] `value` is not a valid prop, do you mean `checked`?',
    );
    errorSpy.mockRestore();
  });

  it('should support data-*、aria-* and custom attribute', () => {
    const wrapper = mount(<Checkbox data-text="123" aria-hidden="true" />);
    expect(wrapper.render()).toMatchSnapshot();
  });
});
