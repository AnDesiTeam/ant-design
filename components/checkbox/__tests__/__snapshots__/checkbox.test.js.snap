// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox rtl render component should be rendered correctly in RTL direction 1`] = `
<label
  class="ant-checkbox-wrapper ant-checkbox-rtl"
>
  <span
    class="ant-checkbox"
  >
    <input
      class="ant-checkbox-input"
      type="checkbox"
    />
    <span
      class="ant-checkbox-inner"
    />
  </span>
</label>
`;

exports[`Checkbox should support data-*„ÄÅaria-* and custom attribute 1`] = `
<label
  aria-hidden="true"
  class="ant-checkbox-wrapper"
  data-text="123"
>
  <span
    class="ant-checkbox"
  >
    <input
      aria-hidden="true"
      class="ant-checkbox-input"
      data-text="123"
      type="checkbox"
      value=""
    />
    <span
      class="ant-checkbox-inner"
    />
  </span>
</label>
`;
